File: Shared/Data/WebService.cs
using System;
using System.IO;
using System.Net;
using System.Text;
using System.Linq;
using System.Threading.Tasks;
using System.Collections.Generic;

using Xamarin.SSO.Client;

namespace XamarinStore
{
	public class WebService
	{
		public static readonly WebService Shared = new WebService ();

		public User CurrentUser { get; set; }
		public Order CurrentOrder { get; set; }

		XamarinSSOClient client = new XamarinSSOClient ("https://auth.xamarin.com", "0c833t3w37jq58dj249dt675a465k6b0rz090zl3jpoa9jw8vz7y6awpj5ox0qmb");

		public WebService ()
		{
			CurrentOrder = new Order ();
		}

		public async Task<bool> Login (string username, string password)
		{
			AccountResponse response;
			try {
				var request = Task.Run (() => response = client.CreateToken (username, password));
				response = await request;
				if (response.Success) {
					var user = response.User;
					CurrentUser = new  User {
						LastName = user.LastName,
						FirstName = user.FirstName,
						Email = username,
						Token = response.Token
					};
					return true;
				} else {
					Console.WriteLine ("Login failed: {0}", response.Error);
				}
			} catch (Exception ex) {
				Console.WriteLine ("Login failed for some reason...: {0}", ex.Message);
			}
			return false;
		}

		List<Product> products;
		public async Task<List<Product>> GetProducts()
		{
			if (products == null) {
				products = await Task.Factory.StartNew (() => {
					try {
						string extraParams = "";

						//TODO: Get a Monkey!!!
						//extraParams = "?includeMonkeys=true";

						var request = CreateRequest ("products" + extraParams);

						string response = ReadResponseText (request);
						return Newtonsoft.Json.JsonConvert.DeserializeObject<List<Product>> (response);
					} catch (Exception ex) {
						Console.WriteLine (ex);
						return new List<Product> ();
					}
				});
			}
			return products;
		}
		bool hasPreloadedImages;
		public async Task PreloadImages(float imageWidth)
		{
			if (hasPreloadedImages)
				return;
			hasPreloadedImages = true;
			//Lets precach the countries too
			#pragma warning disable 4014
			GetCountries ();
			#pragma warning restore 4014
			await Task.Factory.StartNew (() => {
				var imagUrls = products.SelectMany (x => x.ImageUrls.Select (y => Product.ImageForSize (y, imageWidth))).ToList ();
				imagUrls.ForEach( async (x) => await FileCache.Download(x));

			});


		}
		List<Country> countries = new List<Country>();
		public Task<List<Country>> GetCountries()
		{
			return Task.Factory.StartNew (() => {
				try {

					if(countries.Count > 0)
						return countries;

					var request = CreateRequest ("Countries");
					string response = ReadResponseText (request);
					countries = Newtonsoft.Json.JsonConvert.DeserializeObject<List<Country>> (response);
					return countries;
				} catch (Exception ex) {
					Console.WriteLine (ex);
					return new List<Country> ();
				}
			});
		}

		public async Task<string> GetCountryCode(string country)
		{
			var c = (await GetCountries ()).FirstOrDefault (x => x.Name == country) ?? new Country();

			return c.Code;
		}
		//No need to await anything, and no need to spawn a task to return a list.
		#pragma warning disable 1998
		public  async Task<List<string>> GetStates(string country)
		{
			if (country.ToLower () == "united states")
				return new List<string> {
					"Alabama",
					"Alaska", 
					"Arizona",
					"Arkansas",
					"California",
					"Colorado",
					"Connecticut",
					"Delaware",
					"District of Columbia",
					"Florida",
					"Georgia",
					"Hawaii",
					"Idaho",
					"Illinois",
					"Indiana",
					"Iowa",
					"Kansas",
					"Kentucky",
					"Louisiana",
					"Maine",
					"Maryland",
					"Massachusetts",
					"Michigan",
					"Minnesota",
					"Mississippi",
					"Missouri",
					"Montana",
					"Nebraska",
					"Nevada",
					"New Hampshire",
					"New Jersey",
					"New Mexico",
					"New York",
					"North Carolina",
					"North Dakota",
					"Ohio",
					"Oklahoma",
					"Oregon",
					"Pennsylvania",
					"Rhode Island",
					"South Carolina",
					"South Dakota",
					"Tennessee",
					"Texas",
					"Utah",
					"Vermont",
					"Virginia",
					"Washington",
					"West Virginia",
					"Wisconsin",
					"Wyoming",
				};
			return new List<string> ();
		}
		#pragma warning restore 1998

		static HttpWebRequest CreateRequest(string location)
		{
			var request = (HttpWebRequest)WebRequest.Create ("https://xamarin-store-app.xamarin.com/api/"+ location);
			request.Method = "GET";
			request.ContentType = "application/json";
			request.Accept = "application/json";
			return request;
		}
			

		public Task<OrderResult> PlaceOrder (User user, bool verify = false) {
			return Task.Factory.StartNew (() => {
				try {
					var content = Encoding.UTF8.GetBytes (CurrentOrder.GetJson (user));

					var request = CreateRequest ("order" + (verify ? "?verify=1" : ""));
					request.Method = "POST";
					request.ContentLength = content.Length;

					using (Stream s = request.GetRequestStream ()) {
						s.Write (content, 0, content.Length);
					}
					string response = ReadResponseText (request);
					var result = Newtonsoft.Json.JsonConvert.DeserializeObject<OrderResult> (response);
					if(!verify && result.Success)
						CurrentOrder = new Order();
					return result;
				} catch (Exception ex) {
					return new OrderResult {
						Success = false,
						Message = ex.Message,
					};
				}
			});
		}
			

		protected static string ReadResponseText (HttpWebRequest req) {
			using (WebResponse resp = req.GetResponse ()) {
				using (Stream s = (resp).GetResponseStream ()) {
					using (var r = new StreamReader (s, Encoding.UTF8)) {
						return r.ReadToEnd ();
					}
				}
			}
		}
	}
}

File: F#/Shared/WebService.fs
module WebService

    open System
    open System.IO
    open System.Net
    open System.Runtime.Serialization
    open Newtonsoft.Json

    open User
    open Order
    open Helpers
    open Product
    open XamarinSSOClient
    
    [<DataContract;CLIMutable>]
    [<JsonObject(MemberSerialization=MemberSerialization.OptOut)>]
    type Country = 
        { 
            [<DataMember(Name="code")>]
            Code: string
            [<DataMember(Name="name")>]
            Name: string 
        }
    
    let ReadResponseText (req : HttpWebRequest) =
        use response = req.GetResponse ()
        use s = response.GetResponseStream()
        use r = new StreamReader (s, System.Text.Encoding.UTF8)
        r.ReadToEnd()
    
    let CreateRequest location =
        let request = WebRequest.Create ("https://xamarin-store-app.xamarin.com/api/"+ location) :?> HttpWebRequest
        request.Method <- "GET"
        request.ContentType <- "application/json"
        request.Accept <- "application/json"
        request

    let mutable CurrentOrder : Order = Order.CreateOrder()

    type WebService() =
        let client = new XamarinSSOClient ("https://auth.xamarin.com", "0c833t3w37jq58dj249dt675a465k6b0rz090zl3jpoa9jw8vz7y6awpj5ox0qmb")
        let mutable products = [||]
        let mutable hasPreloadedImages = false
        let mutable countries = [||]

        member val CurrentUser = User.CreateUser() with get,set

        member this.Login username password = async {
            try
                let! token = client.CreateToken username password
                match token with
                | Failure msg ->return false
                | Success accountResponse -> if not accountResponse.Success then
                                                 Console.WriteLine ("Login failed: {0}", accountResponse.Error)
                                                 return false
                                             else
                                                 let user = accountResponse.User
                                                 this.CurrentUser <- { FirstName = user.FirstName
                                                                       LastName = user.LastName
                                                                       Email = user.Email
                                                                       Token = accountResponse.Token 
                                                                       Address = ""; Address2 = ""; City = ""; State = ""; ZipCode = ""; Phone = ""; Country = ""}
                                                 return true
            with | e-> Console.WriteLine ("Login failed for some reason...: {0}", e.Message)
                       return false }

        member this.GetProducts ():Async<Product []> = async {
            match products with
            | [||] -> try
                          return! System.Threading.Tasks.Task.Run(fun _->
                              //let extraParams = "&includeMonkeys=true"
                              let extraParams = ""
                              let response = ReadResponseText (CreateRequest ("products?shirttype=fsharp"+extraParams))
                              products <- JsonConvert.DeserializeObject<Product array>(response)
                              products ) |> Async.AwaitTask
                      with | e-> Console.WriteLine e
                                 return [||]
            | _ -> return products }

        member this.GetCountries() = async {
            match countries with
            | [||] -> try
                          return! System.Threading.Tasks.Task.Run(fun _->
                              let response = ReadResponseText (CreateRequest "Countries")
                              countries <- JsonConvert.DeserializeObject<Country array>(response)
                              countries ) |> Async.AwaitTask
                      with | e-> Console.WriteLine e
                                 return [||]
            | _ -> return countries }

        member this.PreloadImages imageWidth = async {
            if hasPreloadedImages then
                ()
            else hasPreloadedImages <- true
                 this.GetCountries() |> Async.Ignore |> Async.StartImmediate
                 do! products |> Array.collect (fun x->x.GetImageUrls() )
                              |> Array.map (fun y->Product.ImageForSize(y, imageWidth) 
                                                   |> FileCache.Download )
                              |> Async.Parallel |> Async.Ignore }

        member this.GetCountryCodeFromName name = async {
            let! countries = this.GetCountries() 
            match countries |> Array.tryFind (fun x-> x.Name = name) with
            | Some(country) -> return country.Code
            | None -> return "" }

        member this.GetCountryNameFromCode code = async {
            let! countries = this.GetCountries() 
            match countries |> Array.tryFind (fun x-> x.Code = code) with
            | Some(country) -> return country.Name
            | None -> return "" }
            
        member this.GetStates (country:String) =
            if country.ToLower() = "united states" then
                [ "Alabama"
                  "Alaska" 
                  "Arizona"
                  "Arkansas"
                  "California"
                  "Colorado"
                  "Connecticut"
                  "Delaware"
                  "District of Columbia"
                  "Florida"
                  "Georgia"
                  "Hawaii"
                  "Idaho"
                  "Illinois"
                  "Indiana"
                  "Iowa"
                  "Kansas"
                  "Kentucky"
                  "Louisiana"
                  "Maine"
                  "Maryland"
                  "Massachusetts"
                  "Michigan"
                  "Minnesota"
                  "Mississippi"
                  "Missouri"
                  "Montana"
                  "Nebraska"
                  "Nevada"
                  "New Hampshire"
                  "New Jersey"
                  "New Mexico"
                  "New York"
                  "North Carolina"
                  "North Dakota"
                  "Ohio"
                  "Oklahoma"
                  "Oregon"
                  "Pennsylvania"
                  "Rhode Island"
                  "South Carolina"
                  "South Dakota"
                  "Tennessee"
                  "Texas"
                  "Utah"
                  "Vermont"
                  "Virginia"
                  "Washington"
                  "West Virginia"
                  "Wisconsin"
                  "Wyoming" ]
            else
                []
        member this.PlaceOrder(user, ?verify) = async {
            try 
                
                return! System.Threading.Tasks.Task.Run(fun _->
                    let verify = if Some(true) = verify then true else false
                    let json = { CurrentOrder with Products = CurrentOrder.Products |> List.map (fun x->x.CreateCopy()) }.GetJson user
                    let content = encoding.GetBytes json
                    let request = CreateRequest ("fsharporder" + if verify then "?verify=1" else "")
                    request.Method <- "POST"
                    request.ContentLength <- int64 content.Length
                     
                    use s = request.GetRequestStream()
                    s.Write(content, 0, content.Length)

                    let response = ReadResponseText request
                    let result = JsonConvert.DeserializeObject<OrderResult> response

                    if (not verify) && result.Success then
                        CurrentOrder <- Order.CreateOrder()

                    result ) |> Async.AwaitTask
            with | e-> return { Success=false; OrderNumber=""; Message=e.Message } }

        member this.ValidateUser user = async {
            if      user.FirstName = "" then return Failure "First name is required"
            else if user.LastName  = "" then return Failure "Last name is required"
            else if user.Phone     = "" then return Failure "Phone number is required"
            else if user.Address   = "" then return Failure "Address is required"
            else if user.City      = "" then return Failure "City is required"
            else if user.Country   = "" then return Failure "Country is required"
            else if user.ZipCode   = "" then return Failure "Zip Code is required"
            else if user.Country.ToLower () = "usa" then 
                let! countryCode = this.GetCountryNameFromCode user.Country
                let states = this.GetStates countryCode
                if states |> List.exists ((=) user.State) then
                    return Success "User is valid"
                else return Failure ("\"" + user.State + "\"" +  " is not a valid state")
            else return Success "User is valid" }
                
    let Shared = new WebService()